lms/djangoapps/courseware/grades.py

```diff
@@ -26,10 +27,97 @@
 from submissions import api as sub_api  # installed from the edx-submissions repository
 from opaque_keys import InvalidKeyError
 
-
 log = logging.getLogger("edx.courseware")
 
 
+class MaxScoresCache(object):
+    """
+    A cache for unweighted max scores for problems.
+
+    The key assumption here is that any problem that has not yet recorded a
+    score for a user is worth the same number of points. An XBlock is free to
+    score one student at 2/5 and another at 1/3. But a problem that has never
+    issued a score -- say a problem two students have only seen mentioned in
+    their progress pages and never interacted with -- should be worth the same
+    number of points for everyone.
+    """
+    def __init__(self, cache_prefix):
+        self.cache_prefix = cache_prefix
+        self._max_scores_cache = {}
+        self._max_scores_updates = {}
+
+    @classmethod
+    def create_for_course(cls, course):
+        return cls(course.subtree_edited_on.isoformat())
+
+    def fetch_from_remote(self, locations):
+        """
```

Comment:
I don't know honestly. I thought that having it separate could make testing easier and be a parallel to the pushing code, but maybe that's silly.